### ISSUE TITLE
Trait bound is not satisfied if two errors implement the same #[from]

### ISSUE THREAD

**Issue by GCastilho** [tmyers273/error-context#2]
So, I don't know if I'm doing something wrong, but if I have two errors that implement the same `#[from]` trait I get a compilation error and can't use neither

Here's an example:

```rust
#[derive(Debug, Error)]
enum InnerA {
    #[error("InnerA Sqlx")]
    Sqlx(#[from] sqlx::Error),
}

impl_context!(OuterA(InnerA));

#[derive(Debug, Error)]
enum InnerB {
    #[error("InnerB Sqlx")]
    Sqlx(#[from] sqlx::Error),
}

impl_context!(OuterB(InnerB));

async fn outer_a(pool: &mut Pool<Sqlite>) -> Result<(), OuterA> {
    let mut tx = pool.begin().await?;
    sqlx::query!("CREATE TABLE IF NOT EXISTS some_table (id INTEGER PRIMARY KEY AUTOINCREMENT)")
        .execute(&mut *tx)
        .await
        .context("outer_a")?; // <---- the trait bound `std::result::Result<SqliteQueryResult, sqlx::Error>: thiserror_context::Context<_, (), _>` is not satisfied

    Ok(())
}
```

![image](https://github.com/user-attachments/assets/c20439ae-feab-472f-b3ef-7e66ae10005d)

The error occur even if they are in different modules

I've made a repository to reproduce it: https://github.com/GCastilho/thiserror-context-trait-not-satisfied-reproduction. Instructions on README

Please let me know if I can help with anything. Thank you

**Comment by GCastilho** [comment]
I've implemented a test case on a [branch of mine](https://github.com/GCastilho/error-context/tree/fix/multiple_errors_same_from) that reproduces this build error in an even simpler way

**Comment by tmyers273** [comment]
Hey @GCastilho,

Thanks for the detailed bug report!

Looking at your branch, it looks like the anyhow import may be causing some issues there. [Link](https://github.com/tmyers273/error-context/compare/main...GCastilho:error-context:fix/multiple_errors_same_from#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R393). The latest [commit](https://github.com/tmyers273/error-context/commit/c82f25ce9a20854bd1dd79a49cff33a5668b5ab8#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R446) seems to be working with that particular test case.

However, I was able to reproduce the original sqlx version you posted. 

TLDR: Adding an explicit conversion should get things compiling. I know it's not the most ergonomic, but I wasn't able to figure out a way to avoid it in a case like this.

```rust
async fn outer_a(pool: &mut Pool<Sqlite>) -> Result<(), OuterA> {
    let mut tx = pool.begin().await?;
    sqlx::query!(
        "CREATE TABLE IF NOT EXISTS some_table (id INTEGER PRIMARY KEY AUTOINCREMENT)"
    )
    .execute(&mut *tx)
    .await
    // .map_err(InnerA::from) // <---- uncomment this line
    .context("outer_a")?; // <---- the trait bound `std::result::Result<SqliteQueryResult, sqlx::Error>: thiserror_context::Context<_, (), _>` is not satisfied

    Ok(())
}
```

It looks like there were 2 factors contributing to the error:
1. the sqlx response was `Result<DB::QueryResult, _>`, while the function response was `Result<(), _>`. That caused the Ok variant type of each Result to be differen, hiding the real error.
1. the compiler can't infer if the sqlx error should be converted to `InnerA` or `InnerB`.

**Comment by GCastilho** [comment]
It seems that using the `.map_err(InnerA::from)` is the only way to make the code to compile

I've investigated a little further and the problem lies in the `Into<$out>` part
![image](https://github.com/user-attachments/assets/c5ff5a96-700d-46b4-b016-591c6b7f1817)
If there are multiple `Into<$out>` implementations then the compiler can't know if you're referring to `InnerA` or `InnerB` (as you mentioned)

This seems to be caused by the `?` not being smart enough to understand that the return type, `OuterA`, implies that `InnerA` is the only valid implementation for `Context` here

I've update my [branch](https://github.com/GCastilho/error-context/tree/fix/multiple_errors_same_from) with a modified test case where the error occur, for completion sake,
![image](https://github.com/user-attachments/assets/de0f6a56-9b36-4d6f-aa7e-c71a2a519242)
but I don't think there is anything the lib can do to fix this without an explicit conversion

Thanks for the quick patch, I managed to solve my original problem with it and the explicit conversion

I think the issue can be closed too
